sbi/inference/snpe/snpe_c.py

```diff
@@ -231,5 +370,354 @@ def _log_prob_proposal_posterior(
 
         return log_prob_proposal_posterior
 
-    def _log_prob_proposal_MoG(self, theta: Tensor, x: Tensor, masks: Tensor) -> Tensor:
-        raise NotImplementedError
+    def _log_prob_proposal_posterior_mog(
+        self, theta: Tensor, x: Tensor, proposal: DirectPosterior,
+    ) -> Tensor:
+        """
+        Return log-probability of the proposal posterior for MoG proposal.
+
+        For MoG proposals and MoG density estimators, this can be done in closed form
+        and does not require atomic loss (i.e. there will be no leakage issues).
+
+        Notation:
+
+        m are mean vectors.
+        prec are precision matrices.
+        cov are covariance matrices.
+
+        _p at the end indicates that it is the proposal.
+        _d indicates that it is the density estimator.
+        _pp indicates the proposal posterior.
+
+        All tensors will have shapes (batch_dim, num_components, ...)
+
+        Args:
+            theta: Batch of parameters θ.
+            x: Batch of data.
+            proposal: Proposal distribution.
+
+        Returns:
+            Log-probability of the proposal posterior.
+        """
+
+        # Evaluate the proposal.
```

Comment:
add comment why we have to dig into the MDN code here - to get access to the component leights (logits).